keps/sig-api-machinery/00xx-publish-crd-openapi.md

```diff
@@ -0,0 +1,142 @@
+---
+kep-number: 0
+title: Publish CRD OpenAPI
+authors:
+  - "@roycaihw"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@liggitt"
+  - "@mbohlool"
+  - "@sttts"
+approvers:
+  - "@smarterclayton"
+editor: TBD
+creation-date: 2019-02-07
+last-updated: 2019-02-07
+status: provisional
+see-also:
+  - [Validation for CustomResources design doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md)
+---
+
+# Publish CRD OpenAPI
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+In CustomResourceDefinition (CRD) we allow CRD author to define JSON-Schema, to
+enable server-side [validation for CustomResources (CR)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md).
+The JSON-Schema format is compatible for creating OpenAPI documentation for CRs,
+which can be used by clients like kubectl to perform client-side validation,
+schema explanation (`kubectl explain`), and client generation.
+This KEP purposes using the JSON-Schema to create and publish OpenAPI
+documentation for CRs.
+
+## Motivation
+
+Publishing CRD OpenAPI enables client-side validation, schema explanation and
+client generation for CRs. It covers the gap between CRD and native Kubernetes
+APIs, which already support OpenAPI documentation.
+
+Publishing CRD OpenAPI is also noted as potential followup in [Validation for CustomResources Implementation Plan](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md#implementation-plan).
+
+### Goals
+
+* For every CRD that we serve, publish Paths (operations that we support on
+  resource and subresources) and Definitions (for both CR object and CR list
+  object) in OpenAPI documentation to fully demonstrate the existence of the
+  API.
+* For CRDs with validation JSON-Schema defined, the CR object Definition should
+  include both CRD validation JSON-Schema and native Kubernetes ObjectMeta and
+  TypeMeta properties.
+* For CRDs without validation JSON-Schema, the CR object Definition will be an
+  emtpy object which validate arbitrary request on client-side.
```

Comment:
rephrase/simplify to "For CRDs without validation JSON-Schema, the CRD object definition will be as complete as possible while still maintaining compatibility with the openapi spec and with supported kubernetes components"